server/datastore/mysql/software.go

```diff
@@ -640,88 +646,122 @@ func (ds *Datastore) SoftwareByID(ctx context.Context, id uint) (*fleet.Software
 // After aggregation, it cleans up unused software (e.g. software installed
 // on removed hosts, software uninstalled on hosts, etc.)
 func (ds *Datastore) CalculateHostsPerSoftware(ctx context.Context, updatedAt time.Time) error {
-	resetStmt := `
-    UPDATE software_host_counts
-    SET hosts_count = 0, updated_at = ?`
-
-	queryStmt := `
-    SELECT count(*), software_id
-    FROM host_software
-    WHERE software_id > 0
-    GROUP BY software_id`
-
-	insertStmt := `
-    INSERT INTO software_host_counts
-      (software_id, hosts_count, updated_at)
-    VALUES
-      %s
-    ON DUPLICATE KEY UPDATE
-      hosts_count = VALUES(hosts_count),
-      updated_at = VALUES(updated_at)`
-	valuesPart := `(?, ?, ?),`
+	const (
+		resetStmt = `
+      UPDATE software_host_counts
+      SET hosts_count = 0, updated_at = ?`
+
+		// team_id is added to the select list to have the same structure as
+		// the teamCountsStmt, making it easier to use a common implementation
+		globalCountsStmt = `
+      SELECT count(*), 0 as team_id, software_id
+      FROM host_software
+      WHERE software_id > 0
+      GROUP BY software_id`
+
+		teamCountsStmt = `
+      SELECT count(*), h.team_id, hs.software_id
+      FROM host_software hs
+      INNER JOIN hosts h
+      ON hs.host_id = h.id
+      WHERE h.team_id IS NOT NULL AND hs.software_id > 0
+      GROUP BY hs.software_id, h.team_id`
+
+		insertStmt = `
+      INSERT INTO software_host_counts
+        (software_id, hosts_count, team_id, updated_at)
+      VALUES
+        %s
+      ON DUPLICATE KEY UPDATE
+        hosts_count = VALUES(hosts_count),
+        updated_at = VALUES(updated_at)`
+
+		valuesPart = `(?, ?, ?, ?),`
+
+		cleanupSoftwareStmt = `
+      DELETE s
+      FROM software s
+      LEFT JOIN software_host_counts shc
+      ON s.id = shc.software_id
+      WHERE
+        shc.software_id IS NULL OR
+        (shc.team_id = 0 AND shc.hosts_count = 0)`
```

Comment:
What about the deletion of `software` rows where `shc.team_id != 0` and `shc.hosts_count = 0`?